assets/app/features/upgrade/controllers/upgrade-nodes.controller.js

```diff
@@ -100,6 +150,35 @@
             vm.nodesUpgrade.totalNodes = response.data.upgraded_nodes + response.data.remaining_nodes;
             vm.nodesUpgrade.currentNodes = response.data.current_nodes;
             vm.nodesUpgrade.currentAction = response.data.current_node_action;
+            vm.nodesUpgrade.computeNodesPostponedFlag = response.data.compute_nodes_postponed;
+
+            // check for the case of postpone upgrade of compute nodes, after controller nodes were
+            // upgraded, step status would still be 'running' while substep status would be 'finished'
+            vm.nodesUpgrade.subStep = response.data.current_substep;
+            vm.nodesUpgrade.subStepStatus = response.data.current_substep_status;
+            if (vm.nodesUpgrade.subStep === 'controller_nodes' && vm.nodesUpgrade.subStepStatus === 'finished'
+                && response.data.nodes_selected_for_upgrade === 'controllers') {
+                vm.nodesUpgrade.running = false;
+
+                // show partially upgrade message
+                vm.nodesUpgrade.upgradeControllersDone = true;
```

Comment:
`upgradeStepsFactory.getUpgradeStep() === NODE_UPGRADE_STEP.compute` seems to match `vm.nodesUpgrade.upgradeControllersDone`. You can use the value from `upgradeStepsFactory` everywhere and remove `upgradeControllersDone`.

Reply:
vm.nodesUpgrade.upgradeControllersDone is used in upgrade-nodes-page.jade. I'm not sure how to use upgradeStepsFactory in upgrade-nodes-page.jade. If you can point me to an example of how to do that, I'd be happy to make the switch.

Reply:
You can expose them via `vm` like we do here: https://github.com/crowbar/crowbar-ui/blob/master/assets/app/features/upgrade/upgrade.controller.js#L32

Reply:
I don't see that exposing the upgradeStepsFactory call via vm will save us any time in this case because we already have upgradeControllersDone from vm. Plus upgradeControllersDone reads easier in upgrade-nodes-page.jade, comparing to getUpgradeStep() === 'compute'. But I'll change it in order to get this code closer to being merged.

Reply:
It's not about performance but rather about keeping the state in one place. "Single source of truth" is always beneficial when you have lots of different variables. If you have two independent variables which are always changed together then in the future someone can forget to set one of them in another place which will lead to undesired results.
If you want, you can add some convenience functions to avoid the comparison. The point is that there should be only one "state" variable which holds given piece of information.